sentry-ruby/lib/sentry/open_telemetry/span_processor.rb

```diff
@@ -0,0 +1,63 @@
+module Sentry
+  module OpenTelemetry
+    class SpanProcessor
+      def initialize
+        @otel_span_map = {}
+      end
+
+      def on_start(otel_span, _parent_context)
+        return if from_sentry_sdk?(otel_span)
+
+        scope = Sentry.get_current_scope
+        parent_sentry_span = scope.get_span
+
+        sentry_span = if parent_sentry_span
+          parent_sentry_span.start_child(op: otel_span.name)
+        else
+          options = { name: otel_span.name, op: otel_span.name }
+          transaction = Sentry::Transaction.from_sentry_trace(scope.trace, **options) if scope.trace
+          scope.set_trace(nil)
+          Sentry.start_transaction(transaction: transaction, **options)
+        end
+
+        scope.set_span(sentry_span)
+        @otel_span_map[otel_span.context.span_id] = [sentry_span, parent_sentry_span]
+      end
+
+      def on_finish(otel_span)
+        current_scope = Sentry.get_current_scope
+        sentry_span, parent_span = @otel_span_map.delete(otel_span.context.span_id)
+        return unless sentry_span
+
+        sentry_span.set_op(otel_span.name)
+        current_scope.set_transaction_name(otel_span.name) if sentry_span.is_a?(Sentry::Transaction)
+
+        otel_span.attributes&.each do |key, value|
+          sentry_span.set_data(key, value)
+          if key == "db.statement"
+            sentry_span.set_description(value)
+          end
+        end
+
+        sentry_span.finish
+        current_scope.set_span(parent_span) if parent_span
+      end
+
+      def force_flush(timeout: nil)
+        # no-op: we rely on Sentry.close being called for the same reason as
+        # whatever triggered this shutdown.
+      end
+
+      def shutdown(timeout: nil)
+        # no-op: we rely on Sentry.close being called for the same reason as
+        # whatever triggered this shutdown.
+      end
+
+      private
+
+      def from_sentry_sdk?(otel_span)
+        caller.any? { |line| line =~ /lib[\\\/]sentry[\\\/]background_worker.rb/ }
```

Comment:
This is here mostly to stop transactions from being made out of the HTTP spans that get created when the background worker sends traces to Sentry. I think `caller` is expensive. We could do this from properties of the OTel span instead, but note that `HTTP CONNECT` spans don't have URL info (so we can't check against the dsn like we do in the `Net::HTTP` monkey patch).

Reply:
`caller` [allocates a new string for each frame on the stack](https://github.com/ruby/ruby/blob/58c8b6e86273ccb7a1b903d9ab35956b69b3b1bf/vm_backtrace.c#L402-L430). And in Rails projects, it's easy to have 100+ frames because all the frameworks/libraries loaded. So using `caller` here will have a huge performance impact.

Reply:
Thank you for the source! That's what I feared. Rather than trying to detect whether it's a Sentry span, we could perhaps instead prevent HTTP client spans from turning into transactions. For the majority of use cases I imagine you would only really want to collect them as part of a larger transaction (something associated with a framework).

Reply:
I would be fine with just skipping HTTP CONNECT in general.

Reply:
this has been solved in https://github.com/getsentry/sentry-ruby/pull/1876/commits/26127e5244fcbf377152c0054fd56c02dacd2287#diff-a614943cb316a2889baab64fd91e7d69bfe10bbff44717fcf985655d16988d17R75-R89

basically we can always check `net.peer.name` set [here for CONNECT](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/ea896ed70cc6f6ce1b674c89376c90dfd70f769d/instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb#L57) and [here for the others](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/ea896ed70cc6f6ce1b674c89376c90dfd70f769d/instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb#L27) against our dsn so it seems to work.